src/Functions/GatherUtils/Algorithms.h

```diff
@@ -418,15 +433,92 @@ bool sliceHasImpl(const FirstSliceType & first, const SecondSliceType & second,
                 has = true;
         }
 
-        if (has && !all)
+        if (has && search_type == ArraySearchType::Any)
             return true;
 
-        if (!has && all)
+        if (!has && search_type == ArraySearchType::All)
             return false;
+    }
+    return search_type == ArraySearchType::All;
+}
+/// Methods to check if first array has elements from second array, overloaded for various combinations of types.
+template < typename FirstSliceType,
+           typename SecondSliceType,
+           bool (*isEqual)(const FirstSliceType &, const SecondSliceType &, size_t, size_t),
+           bool (*isEqualUnary)(const SecondSliceType &, size_t, size_t)>
+bool sliceHasImplSubSeq(const FirstSliceType & first, const SecondSliceType & second, const UInt8 * first_null_map, const UInt8 * second_null_map) noexcept
+{
+    const bool has_first_null_map = first_null_map != nullptr;
+    const bool has_second_null_map = second_null_map != nullptr;
+
+    if (second.size == 0) 
+        return true;
+
+    const auto aux = buildAux<SecondSliceType, isEqualUnary>(second, second_null_map);
```

Comment:
This will produce memory leak, because the result is a raw memory pointer and is never deleted

Reply:
I would also recommend to rename this function to mention the KMP algorithm, so that you can elude writing clarifying comment.